include/llvm-dialects/Dialect/OpMap.h

```diff
@@ -0,0 +1,356 @@
+      bool isOverload =
+          desc.getKind() == OpDescription::Kind::DialectWithOverloads;
+      return Key.first == desc.getMnemonic() && Key.second == isOverload;
+    }
+  };
+
+public:
+  OpMap() = default;
+
+  // -------------------------------------------------------------
+  // Convenience constructor to initialize the OpMap from a set of
+  // OpDescription/Value pairs.
+  // -------------------------------------------------------------
+  OpMap(std::initializer_list<std::pair<OpDescription, ValueT>> vals) {
+    for (const std::pair<OpDescription, ValueT> &val : vals) {
+      insert_or_get(val.first, val.second);
+    }
+  }
+
+  // -------------------------------------------------------------
+  // Comparison operator overloads.
+  // -------------------------------------------------------------
+  bool operator==(const OpMap &rhs) const {
+    return m_coreOpcodes == rhs.m_coreOpcodes &&
+           m_intrinsics == rhs.m_intrinsics && m_dialectOps == rhs.m_dialectOps;
+  }
+
+  bool operator!=(const OpMap &rhs) const { return !(*this == rhs); }
+
+  // -------------------------------------------------------------
+  // contains checks if a given op is contained in any of the
+  // internal map containers.
+  // -------------------------------------------------------------
+
+  // Check if the map contains either a core operation or an intrinsic.
+  bool contains(unsigned op) const {
+    return m_coreOpcodes.contains(op) || m_intrinsics.contains(op);
+  }
+
+  // Check if the map contains an OpDescription created for a given dialect
+  // operation type.
+  template <typename OpT> bool contains() const {
+    static OpDescription desc = OpDescription::get<OpT>();
+    return contains(desc);
+  }
+
+  // Check if the map contains an op described by an OpDescription.
+  bool contains(const OpDescription &desc) const {
+    if (desc.isCoreOp() || desc.isIntrinsic()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (contains(op))
+          return true;
+      }
+    }
+
+    for (const DialectOpKV &dialectOpKV : m_dialectOps) {
+      if (dialectOpKV == desc)
+        return true;
+    }
+
+    return false;
+  }
+
+  // -------------------------------------------------------------
+  // get returns a reference to the value stored in any of the
+  // internal map containers or constructs a new value.
+  // -------------------------------------------------------------
+  template <typename OpT> ValueT &get() {
+    return get(OpDescription::get<OpT>());
+  }
+
+  // Return a reference to the operation described by a given OpDescription.
+  // This includes any of the possible operation kinds.
+  // Constructs a new entry if the operation was found in neither map.
+  ValueT &get(const OpDescription &desc) {
+    if (desc.isCoreOp()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (m_coreOpcodes.contains(op))
+          return m_coreOpcodes[op];
+      }
```

Comment:
I think this loop, which has a dependency on the hash function used in `DenseMap`, proves that the `OpMap` class should refuse `OpDescription`s of multiple core opcodes and intrinsics in most or even all places.

This would also allow you to directly forward this to the correct `DenseMap` instead of forwarding to `insert_or_get` where a redundant kind check happens (which the compiler can probably optimize away if it inlines everything, but still).